pkg/services/ngalert/store/instance_database.go

```diff
@@ -88,30 +91,147 @@ func (st DBstore) ListAlertInstances(ctx context.Context, cmd *models.ListAlertI
 	})
 }
 
-// SaveAlertInstance is a handler for saving a new alert instance.
-func (st DBstore) SaveAlertInstance(ctx context.Context, cmd *models.SaveAlertInstanceCommand) error {
-	return st.SQLStore.WithDbSession(ctx, func(sess *sqlstore.DBSession) error {
-		labelTupleJSON, labelsHash, err := cmd.Labels.StringAndHash()
+// SaveAlertInstances saves all the provided alert instances to the store.
+func (st DBstore) SaveAlertInstances(ctx context.Context, cmd ...models.AlertInstance) error {
+	if !st.FeatureToggles.IsEnabled(featuremgmt.FlagAlertingBigTransactions) {
+		// This mimics the replace code-path by calling SaveAlertInstance in a loop, with a transaction per call.
+		for _, c := range cmd {
+			err := st.SaveAlertInstance(ctx, c)
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	} else {
+		// The function starts a single transaction and batches writes into
+		// statements with `maxRows` instances per statements. This makes for a
+		// fairly efficient transaction without creating statements that are too long
+		// for some databases to process. For example, SQLite has a limit of 999
+		// variables per write.
+		err := st.SQLStore.WithTransactionalDbSession(ctx, func(sess *sqlstore.DBSession) error {
+			keyNames := []string{"rule_org_id", "rule_uid", "labels_hash"}
+			fieldNames := []string{
+				"rule_org_id", "rule_uid", "labels", "labels_hash", "current_state",
+				"current_reason", "current_state_since", "current_state_end", "last_eval_time",
+			}
+			fieldsPerRow := len(fieldNames)
+			maxRows := 20
+			maxArgs := maxRows * fieldsPerRow
+
+			// bigStmt is a prepared statement for a statement that upserts maxRows
+			// into the database. For a large number of instances, it will be used
+			// repeatedly.
+			var bigStmt *core.Stmt
+			defer func() {
+				if bigStmt != nil {
+					closeErr := bigStmt.Close()
+					if closeErr != nil {
+						st.Logger.Warn("failed to close db connection", "err", closeErr)
+					}
+				}
+			}()
+
+			// Prepare our big SQL statement on-demand, instead of initializing up-front
+			getBigStatement := func() (*core.Stmt, error) {
+				if bigStmt != nil {
+					return bigStmt, nil
+				}
+
+				bigUpsertSQL, err := st.SQLStore.Dialect.UpsertMultipleSQL(
+					"alert_instance", keyNames, fieldNames, maxRows)
+				if err != nil {
+					return nil, err
+				}
+
+				bigStmt, err = sess.DB().Prepare(bigUpsertSQL)
```

Comment:
This is just a question as I am not sure - are we actually using the transaction if we are preparing the statement via `*core.DB` rather than via `*xorm.Session`? I'd assume all sessions share the same DB pointer


from xorm:

```
session.go


// DB db return the wrapper of sql.DB
func (session *Session) DB() *core.DB {
	if session.db == nil {
		session.db = session.engine.db
		session.stmtCache = make(map[uint32]*core.Stmt, 0)
	}
	return session.db
}
```


```
engine.go


// NewSession New a session
func (engine *Engine) NewSession() *Session {
	session := &Session{engine: engine}
	session.Init()
	return session
}
```

Reply:
Under the hood, xorm does the same thing when you set `Prepare` on a session. Unfortunately you can't directly execute a sql string with arguments on a xorm session.

```
session.go

func (session *Session) doPrepare(db *core.DB, sqlStr string) (stmt *core.Stmt, err error) {
	crc := crc32.ChecksumIEEE([]byte(sqlStr))
	// TODO try hash(sqlStr+len(sqlStr))
	var has bool
	stmt, has = session.stmtCache[crc]
	if !has {
		stmt, err = db.PrepareContext(session.ctx, sqlStr)
		if err != nil {
			return nil, err
		}
		session.stmtCache[crc] = stmt
	}
	return
}
```

the xorm session holds on to a `tx` object created by `Begin` that it closes in `Commit` or `Rollback`. Unfortunately `xorm` itself can execute raw sql statements but can't repeatedly use a prepared statement. 

If we go with the idea above of using a txn per batch, I'll move to using the xorm `Statement.SQL` method, since it won't benefit us to repeatedly use a prepared statement.

Reply:
> Unfortunately you can't directly execute a prepared statement with arguments on a xorm session.

hmmm - what about this? https://github.com/grafana/grafana/compare/joe/db-close-stmt...xorm-prepared-statements?expand=1

the `exec` goes into a special branch if `sess.Prepare()` had been called:

```
	if session.prepareStmt {
		stmt, err := session.doPrepare(session.DB(), sqlStr)
		if err != nil {
			return nil, err
		}

		res, err := stmt.ExecContext(session.ctx, args...)
		if err != nil {
			return nil, err
		}
		return res, nil
	}
```

Reply:
I went through and read the code again, and I think there just isn't a way to prepare a statement to be executed repeatedly in a xorm session. I was wrong further up-thread and your were right about what connections are shared between method calls.

Starting a transaction in xorm will grab a connection, start and transaction, and release it when the tx closes.
But doing `sess.DB().Prepare` grabs another connection, unrelated to the session's transaction, so this is wasting a whole transactional connection.

I'm working on a rev that breaks this into smaller transactions, which will use a separate statement in each transaction. We can't share those, so there isn't a reason to prepare them at all, and we can use the standard xorm session commands to execute the queries.